### PR TITLE
Add typing for utilities/memory.py

### PR DESCRIPTION
**Pull request by DuYicong515**
## What does this PR do?

Part of Fixes for #7037, #5023. 

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [n/a] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [N/A] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @ananthsub

### CODE REVIEW

**Review by ananthsub — APPROVED** [review]
Thanks for working on this!

File: pytorch_lightning/utilities/memory.py [hunk 1]
```diff
@@ -184,6 +184,6 @@ def get_model_size_mb(model: Module) -> float:
         Number of megabytes in the parameters of the input module.
     """
     model_size = _ByteCounter()
-    torch.save(model.state_dict(), model_size)
+    torch.save(model.state_dict(), cast(IO[bytes], model_size))
```
Comment (carmocca) [inline pytorch_lightning/utilities/memory.py:187]: Did you consider making `_ByteCounter` implement the `BytesIO` interface?

Comment (DuYicong515) [reply pytorch_lightning/utilities/memory.py:187]: Thanks for. the suggestion, I tried this but mypy complains about 
> pytorch_lightning/utilities/memory.py:35: error: Argument 1 of "write" is incompatible with supertype "BufferedIOBase"; supertype defines the argument type as "Union[bytes, Union[bytearray, memoryview, array[Any], mmap, _CData]]"  [override]
pytorch_lightning/utilities/memory.py:35: note: This violates the Liskov substitution principle"
```
class _ByteCounter(BytesIO):
    """Accumulate and stores the total bytes of an object."""

    def __init__(self) -> None:
        self.nbytes: int = 0

    def write(self, data: bytes) -> int:
        """Stores the total bytes of the data."""
        self.nbytes += len(data)
        return self.nbytes

    def flush(self) -> None:
        pass

```
It doesn't like the argument type becoming more specific(bytes v.s. the complex union).  Not sure what's the best way to address this issue (changing the input type of `write` here?) that's why I go with the castings.
Let me know how do you think! Thanks!

Comment (tchaton) [reply pytorch_lightning/utilities/memory.py:187]: @carmocca

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/utilities/memory.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/utilities/memory.py]: I pushed a commit where we use `BytesIO` directly instead of our own class. Let's see if CI agrees.

**Review by carmocca — APPROVED** [review]

**Review by akihironitta — APPROVED** [review]

**Review by Borda — APPROVED** [review]
